src/stactools/ecmwf_forecast/range_codec.py

```diff
@@ -0,0 +1,28 @@
+from numcodecs.abc import Codec
+import numpy as np
+
+class Range(Codec):
+    """Codec providing range compression
+    
+    Parameters
+    ----------
+    """
+    codec_id='range'
+    
+    def _get_start_stop_inc(self, array):
+        start = array[0]
+        stop = array[-1]
+        delta = np.mean(array[1:] - array[0:-1])
+        deltap = np.round(delta,2)
+        if np.abs(delta-deltap) > 1e-10:
+            print("Delta may not be accurate:", delta, deltap)
+        return start, np.round(stop+deltap,2), deltap
+    
+    def encode(self,buf):
+        info = self._get_start_stop_inc(np.frombuffer(buf))
+        info = np.array([*info])
```

Comment:
Potential issue with the dtypes here: if the array we're compressing has integer values, then the start and stop will be ints, but the step will be a float. Creating the ndarray here will make them all floats (possibly causing loss of precision on the ints).


Maybe spend 30 minutes max seeing if you can use a structured array https://numpy.org/doc/stable/user/basics.rec.html to hold the fields with different dtypes. Not worth worrying about too much though, since IIRC we only have floats for this dataset.

Reply:
Updated range_codec.py to take dtype as input in zarr filter.